frontend/subrequests/lint/lint.go

```diff
@@ -124,47 +111,78 @@ func (results *LintResults) ToResult() (*client.Result, error) {
+				return errors.Errorf("sourceIndex points to nil source")
+			}
+		}
+	}
+	return nil
+}
+
 func PrintLintViolations(dt []byte, w io.Writer) error {
-	var warnings LintResults
+	var results LintResults
 
-	if err := json.Unmarshal(dt, &warnings); err != nil {
+	if err := json.Unmarshal(dt, &results); err != nil {
 		return err
 	}
 
-	// Here, we're grouping the warnings by rule name
-	lintWarnings := make(map[string][]Warning)
-	lintWarningRules := []string{}
-	for _, warning := range warnings.Warnings {
-		if _, ok := lintWarnings[warning.RuleName]; !ok {
-			lintWarningRules = append(lintWarningRules, warning.RuleName)
-			lintWarnings[warning.RuleName] = []Warning{}
-		}
-		lintWarnings[warning.RuleName] = append(lintWarnings[warning.RuleName], warning)
+	if err := results.validateWarnings(); err != nil {
+		return err
 	}
-	sort.Strings(lintWarningRules)
-
-	tw := tabwriter.NewWriter(w, 0, 0, 2, ' ', 0)
-	for _, rule := range lintWarningRules {
-		fmt.Fprintf(tw, "Lint Rule %s\n", rule)
-		for _, warning := range lintWarnings[rule] {
-			source := warnings.Sources[warning.Location.SourceIndex]
-			sourceData := bytes.Split(source.Data, []byte("\n"))
-			firstRange := warning.Location.Ranges[0]
-			if firstRange.Start.Line != firstRange.End.Line {
-				fmt.Fprintf(tw, "\t%s:%d-%d\n", source.Filename, firstRange.Start.Line, firstRange.End.Line)
-			} else {
-				fmt.Fprintf(tw, "\t%s:%d\n", source.Filename, firstRange.Start.Line)
-			}
-			fmt.Fprintf(tw, "\t%s\n", warning.Detail)
-			for _, r := range warning.Location.Ranges {
-				for i := r.Start.Line; i <= r.End.Line; i++ {
-					fmt.Fprintf(tw, "\t%d\t|\t%s\n", i, sourceData[i-1])
-				}
-			}
-			fmt.Fprintln(tw)
+
+	sort.Slice(results.Warnings, func(i, j int) bool {
+		warningI := results.Warnings[i]
+		warningJ := results.Warnings[j]
+		sourceIndexI := warningI.Location.SourceIndex
+		sourceIndexJ := warningJ.Location.SourceIndex
+		if sourceIndexI < 0 && sourceIndexJ < 0 {
+			return warningI.RuleName < warningJ.RuleName
+		} else if sourceIndexI < 0 || sourceIndexJ < 0 {
+			return sourceIndexI < 0
+		}
+
+		sourceInfoI := results.Sources[warningI.Location.SourceIndex]
+		sourceInfoJ := results.Sources[warningJ.Location.SourceIndex]
+		if sourceInfoI.Filename != sourceInfoJ.Filename {
+			return sourceInfoI.Filename < sourceInfoJ.Filename
+		}
+		if len(warningI.Location.Ranges) == 0 && len(warningJ.Location.Ranges) == 0 {
+			return warningI.RuleName < warningJ.RuleName
+		} else if len(warningI.Location.Ranges) == 0 || len(warningJ.Location.Ranges) == 0 {
+			return len(warningI.Location.Ranges) == 0
+		}
+
+		return warningI.Location.Ranges[0].Start.Line < warningJ.Location.Ranges[0].Start.Line
+	})
+
+	for _, warning := range results.Warnings {
+		fmt.Fprintf(w, "\n- %s\n%s\n", warning.Detail, warning.Description)
+		if warning.URL != "" {
+			fmt.Fprintf(w, "URL: %s\n", warning.URL)
+		}
+		sourceInfo := results.Sources[warning.Location.SourceIndex]
```

Comment:
This also needs early exit on negative value.

Reply:
👍